client/src/main/java/org/spine3/validate/ValidatingBuilder.java

```diff
@@ -40,16 +52,117 @@
+     *
+     * @param message the message to merge into the state this builder
+     * @return the builder instance
+     */
+    ValidatingBuilder<T, B> mergeFrom(T message);
+
+    /**
+     * Sets an original state for this builder to allow building the new value on top of some
+     * pre-defined value.
+     *
+     * <p>The validation of the passed {@code state} is NOT performed.
+     *
+     * @param state the new state
+     */
+    @Internal
+    void setOriginalState(T state);
+
+    /**
+     * Determines if the current message state has been modified comparing to its original state.
+     *
+     * @return {@code true} if it is modified, {@code false} otherwise
+     */
+    @Internal
+    boolean isDirty();
+
+    /**
+     * Clears the state of this builder.
+     *
+     * <p>In particular, the state of the message being built is reset to the default.
+     * Also the original state (if it {@linkplain #setOriginalState(Message) has been set}
+     * previously) is cleared.
      */
-    T build();
+    @Internal
+    void clear();
+
+    /**
+     * Enumeration of generic type parameters of this interface.
+     */
+    enum GenericParameter {
+
+        /**
+         * The index of the declaration of the generic parameter type {@code <T>}
+         * in {@link ValidatingBuilder}.
+         */
+        MESSAGE(0),
+
+        /**
+         * The index of the declaration of the generic parameter type {@code <B>}
+         * in {@link ValidatingBuilder}
+         */
+        MESSAGE_BUILDER(1);
+
+        private final int index;
+
+        GenericParameter(int index) {
+            this.index = index;
+        }
+
+        public int getIndex() {
+            return this.index;
+        }
+    }
+
+    /**
+     * Provides type information on classes implementing {@link ValidatingBuilder}.
+     */
+    class TypeInfo {
+
+        private TypeInfo() {
+            // Prevent construction from outside.
+        }
+
+        /**
+         * Retrieves the state class of the passed entity class.
+         *
+         * @param builderClass the builder class to inspect
+         * @param <T>          the state type
+         * @return the entity state class
+         */
+        public static <T extends Message> Class<T> getMessageClass(
+                Class<? extends ValidatingBuilder> builderClass) {
+            final Class<T> result =
+                    getGenericParameterType(builderClass,
+                                            GenericParameter.MESSAGE.getIndex());
+            return result;
+        }
+
+        // as the method names are the same, but methods are different.
+        @SuppressWarnings("DuplicateStringLiteralInspection")
+        public static Method getNewBuilderMethod(Class<? extends ValidatingBuilder<?, ?>> cls) {
+            try {
+                return cls.getMethod("newBuilder");
+            } catch (NoSuchMethodException e) {
+                throw illegalArgumentWithCauseOf(e);
```

Comment:
This line is actually tested, though codecov thinks it's not. See `ValidatingBuilderShould.throw_exception_if_accessing_newBuilder_method_for_improperly_declared_builder`.